lib/url.js

```diff
@@ -85,6 +87,22 @@ function urlParse(url, parseQueryString, slashesDenoteHost) {
   return u;
 }
 
+// Browsers encode these in hostnames. Let's encode them too.
+const hostEscapeSpecial = {
+  '!': '%21',
+  '*': '%2A',
+  '(': '%28',
+  ')': '%29'
+  // '\'': '%27'
+};
+
+function encodeHost(host) {
+  host = encodeURIComponent(host);
+  return host.replace(/[!*()]/g, function(char) {
```

Comment:
What about pulling out this regexp and instead building it dynamically (once) from `Object.keys(hostEscapeSpecial)` and then re-using that same instance inside `encodeHost()`?